src/backend/distributed/utils/replication_origin_session_utils.c

```diff
@@ -0,0 +1,222 @@
+		{
+			replorigin_session_reset();
+			replorigin_session_origin = InvalidRepOriginId;
+		}
+	}
+	else
+	{
+		/*Reset Replication Origin in remote session */
+		StringInfo replicationOriginSessionResetQuery = makeStringInfo();
+		appendStringInfo(replicationOriginSessionResetQuery,
+						 "select pg_catalog.pg_replication_origin_session_reset()");
+		ExecuteCriticalRemoteCommand(connection,
+									 replicationOriginSessionResetQuery->data);
+	}
+}
+
+
+/* isReplicationOriginSessionSetup checks if the replication origin is setup
+ * already in the local or remote session.
+ */
+static bool
+isReplicationOriginSessionSetup(MultiConnection *connection)
+{
+	bool result = false;
+	if (connection == NULL)
+	{
+		return replorigin_session_origin != InvalidRepOriginId;
+	}
+	else
+	{
+		/*Setup Replication Origin in remote session */
+		StringInfo isReplicationOriginSessionSetupQuery = makeStringInfo();
+		appendStringInfo(isReplicationOriginSessionSetupQuery,
+						 "SELECT pg_catalog.pg_replication_origin_session_is_setup()");
+		result =
+			ExecuteRemoteCommandAndCheckResult(connection,
+											   isReplicationOriginSessionSetupQuery->data,
+											   "t");
+	}
+	return result;
+}
+
+
+/* isReplicationOriginCreated checks if the replication origin is created
+ * in the local or remote session.*/
+static bool
+isReplicationOriginCreated(MultiConnection *connection, char *originName,
+						   RepOriginId *originId)
+{
+	bool result = false;
+	if (connection == NULL)
+	{
+		*originId = replorigin_by_name(originName, true);
+		result = (*originId != InvalidRepOriginId);
+	}
+	else
+	{
+		/*Setup Replication Origin in remote session */
+		StringInfo isReplicationOriginSessionSetupQuery = makeStringInfo();
+		appendStringInfo(isReplicationOriginSessionSetupQuery,
+						 "SELECT pg_catalog.pg_replication_origin_oid('%s');",
+						 originName);
+
+		/* If the replication origin was already created the above command
+		 * will return the id of the entry in pg_replication_origin table.
+		 * So to check if the entry is there alreay, the retuen value should
+		 * be non-empty and the condition below checks that. */
+		result = !ExecuteRemoteCommandAndCheckResult(connection,
+													 isReplicationOriginSessionSetupQuery
+													 ->data, "");
+	}
+	return result;
+}
+
+
+/* ReplicationOriginSessionCreate creates a new replication origin if it does
+ * not already exist already. To make the replication origin name unique
+ * for different nodes, origin node's id is appended to the prefix citus_internal_.*/
+static RepOriginId
+ReplicationOriginSessionCreate(MultiConnection *connection, char *originName)
+{
+	RepOriginId originId = InvalidRepOriginId;
+	if (connection == NULL)
+	{
+		originId = replorigin_create(originName);
+	}
+	else
+	{
+		StringInfo replicationOriginCreateQuery = makeStringInfo();
+		appendStringInfo(replicationOriginCreateQuery,
+						 "select pg_catalog.pg_replication_origin_create('%s')",
+						 originName);
```

Comment:
nicer to use quote_literal_cstr, since quotes in the format string usually points to an injection bug.